From 7211b2afb492a48c01a88e4fe3c2ce4ea1f7a558 Mon Sep 17 00:00:00 2001
From: Robert Beckett <bob.beckett@collabora.com>
Date: Mon, 3 Nov 2025 17:30:45 +0000
Subject: [PATCH 059/161] HID: multitouch: override focaltech fts3528 phys
 dimensions

Fix false physical dimensions reported by FTS3528 hid reports.
With these corrected, userland can more easily correlate ts to display.

Signed-off-by: Robert Beckett <bob.beckett@collabora.com>
---
 drivers/hid/hid-ids.h        |  3 +++
 drivers/hid/hid-multitouch.c | 39 ++++++++++++++++++++++++++++++++++++
 2 files changed, 42 insertions(+)

diff --git a/drivers/hid/hid-ids.h b/drivers/hid/hid-ids.h
index cb7f16d75..94120a687 100644
--- a/drivers/hid/hid-ids.h
+++ b/drivers/hid/hid-ids.h
@@ -1682,4 +1682,7 @@
 #define USB_VENDOR_ID_HUAWEI			0x12d1
 #define USB_DEVICE_ID_HUAWEI_CD30KBD		0x109b
 
+#define I2C_VENDOR_ID_FOCALTECH		0x2808
+#define I2C_DEVICE_ID_FOCALTECH_FTS3528	0x1015
+
 #endif
diff --git a/drivers/hid/hid-multitouch.c b/drivers/hid/hid-multitouch.c
index 4e19a0c4d..94d0c4fb2 100644
--- a/drivers/hid/hid-multitouch.c
+++ b/drivers/hid/hid-multitouch.c
@@ -1731,6 +1731,28 @@ static void mt_yogabook9_fixup(struct hid_device *hdev, __u8 *rdesc,
 	}
 }
 
+static void mt_report_replace(struct device *dev, __u8 *rdesc, size_t rsize,
+			     const u8 *confirm, const u8 *replace,
+			     size_t pattern_len, const size_t *offsets,
+			     unsigned int num_offsets)
+{
+	size_t offset;
+	int i;
+
+	for (i = 0; i < num_offsets; i++) {
+		offset = offsets[i];
+		if (offset + pattern_len >= rsize) {
+			dev_err(dev, "Failed to fixup hid report: invalid length\n");
+			break;
+		}
+		if (memcmp(rdesc + offset, confirm, pattern_len)) {
+			dev_err(dev, "Failed to fixup hid report: failed to validate: %d\n", i);
+			break;
+		}
+		memcpy(rdesc + offset, replace, pattern_len);
+	}
+}
+
 static const __u8 *mt_report_fixup(struct hid_device *hdev, __u8 *rdesc,
 			     unsigned int *size)
 {
@@ -1764,6 +1786,23 @@ got: %x\n",
 	    hdev->product == USB_DEVICE_ID_LENOVO_YOGABOOK9I)
 		mt_yogabook9_fixup(hdev, rdesc, size);
 
+	if (hdev->vendor == I2C_VENDOR_ID_FOCALTECH &&
+	    hdev->product == I2C_DEVICE_ID_FOCALTECH_FTS3528) {
+		static u8 x_replace[] = {0x46, 0x0b, 0x0a}; /* 2571 *10^-2 cm */
+		static u8 x_with[] = {0x46, 0x40, 0x06}; /* 1600 * 10^-2 cm */
+		static size_t x_offsets[] = {0x37, 0x92, 0xed, 0x148};
+		static u8 y_replace[] = {0x46, 0xaa, 0x05}; /* 1450 *10^-2 cm */
+		static u8 y_with[] = {0x46, 0xe8, 0x03}; /* 1000 * 10^-2 cm */
+		static size_t y_offsets[] = {0x42, 0x9d, 0xf8, 0x153};
+
+		mt_report_replace(&hdev->dev, rdesc, *size, x_replace, x_with,
+				  ARRAY_SIZE(x_replace), x_offsets,
+				  ARRAY_SIZE(x_offsets));
+		mt_report_replace(&hdev->dev, rdesc, *size, y_replace, y_with,
+				  ARRAY_SIZE(y_replace), y_offsets,
+				  ARRAY_SIZE(y_offsets));
+	}
+
 	return rdesc;
 }
 
-- 
2.56.0

